### ISSUE TITLE
Some R code cannot be executed normally

### ISSUE THREAD

**Issue by DmdCalvinChen** [posit-dev/positron#6163]
This s a normal ANN model training code, but the cv_model code cannot execute properly—it gets stuck with no response, and this issue can be reproduced every time. However, it runs normally on RStudio Server. I'm not sure where the scheduling problem lies.

# 引用包
library(neuralnet)
library(NeuralNetTools)
library(caret)
library(pROC)

library(foreach)
library(doParallel)

load("ann_arg.rda")
exp_ann <- exp_vein

  set.seed(20240118)

  # 获取样本数量
  n_samples <- ncol(exp_ann)

  # 创建训练集索引(70%的样本)
  train_index <- sample(1:n_samples, size = round(0.7 * n_samples))

  train_group_list <- group_list[train_index]
  test_group_list <- group_list[-train_index]

  traindata <- exp_ann[, train_index]
  testdata <- exp_ann[, -train_index]

  k <- traindata

  k <- as.data.frame(t(k))
  fen <- as.data.frame(train_group_list)
  # 根据fen中group_list这一列的变量名，新生成一列，将对应healthy的标记为0，对应disease的标记为1
  fen$lasso <- ifelse(fen$train_group_list == "healthy", 0, 1)

  k <- cbind(fen$lasso, k)
  colnames(k)[1] <- "group"
  data <- k

  head(data)

  # 提取分组信息
  groups <- data$group
  # 移除group列，只保留基因表达数据
  gene_expression <- data[, -which(names(data) == "group")]

  # 将分组信息和基因表达数据合并
  # 确保group不在gene_expression中
  gene_expression <- data[, -which(names(data) == "group")]

  # 创建一个新的数据框，其中group作为因子
  model_data <- data.frame(group = as.factor(groups), gene_expression)

  # 检查数据结构
  str(model_data)

  # 定义神经网络公式
  formula <- as.formula(paste("group ~", paste(colnames(model_data)[colnames(model_data) != "group"], collapse = " + ")))

  # 构建神经网络模型
  # 使用L2正则化训练神经网络
  ann_model <- neuralnet(
    formula,
    data = model_data,
    hidden = c(10, 5),
    linear.output = FALSE,
    err.fct = "ce", # 交叉熵损失函数
    lifesign = "full",
    rep = 1,
    algorithm = "rprop+", # 可以使用rprop+优化算法
    stepmax = 1e+08,
  )

  #########################################

  # 如果levels是"0"和"1"，我们需要将它们改为有效的R变量名
  model_data$group <- factor(model_data$group, levels = c("0", "1"), labels = c("Healthy", "Disease"))

  #######################################
  # 2. 使用caret包进行5折交叉验证
  # 确保结果可重复
  train_control <- trainControl(
    method = "cv",
    number = 5,
    classProbs = TRUE,
    summaryFunction = twoClassSummary
  )

  # 使用train函数进行交叉验证
  cv_model <- train(
    formula,
    data = model_data,
    method = "nnet", # 使用神经网络
    trControl = train_control,
    metric = "ROC",
    tuneLength = 10 # 尝试10种不同的参数组合
  )

It gets stuck indefinitely at this step.

**Comment by timtmok** [comment]
Thank you for opening an issue. Your feedback will be helpful!

How large is this dataset? Is it expected to complete quickly and how long does it take when running on RStudio Server?

**Comment by DmdCalvinChen** [comment]
@timtmok 谢谢回复，数据集并不大，内存占用相对于服务器内存量几乎可以忽略不计，rstudio server只需要20s就能完成。重点是，这一步执行的时候会在控制台弹出执行信息，但是positron并不会弹出执行信息，完全处于卡顿状态

**Comment by DmdCalvinChen** [comment]
@timtmok 
[ann_arg.zip] (https://github.com/user-attachments/files/18631081/ann_arg.zip)
This is the data of this ann_arg.rda, through which I can reproduce it on the server.

**Comment by seeM** [comment]
@DmdCalvinChen thank you for the extra info! Could you please share the R kernel debug logs? That would help us investigate further.

1. Set the `positron.r.kernel.logLevel` setting to "debug": 

    <img width="679" alt="Image" src="https://github.com/user-attachments/assets/c6286e15-20d2-46a8-b4a0-69a3e64fcf8f" />
2. Restart Positron.
3. Rerun the script.
4. Run the "Interpreter: Show interpreter output" command.
5. Copy and paste the results here (you can wrap it in `<details>...</details>` to make the text collapsible), or upload it as a file if it's too large.

**Comment by juliasilge** [comment]
@DmdCalvinChen can you try out some simpler code to see if the problem is using doParallel with a remote SSH session. For example, can you execute this with your setup?

```r
library(doParallel)
registerDoParallel(cores=2)
foreach(i=1:3) %dopar% sqrt(i)
```

If that works fine, can you run a simpler caret example, to see if the problem is related to that? For example, can you execute this with your setup?

```r
library(caret)
data(iris)
TrainData <- iris[,1:4]
TrainClasses <- iris[,5]

knnFit1 <- train(TrainData, TrainClasses,
                 method = "knn",
                 preProcess = c("center", "scale"),
                 tuneLength = 10,
                 trControl = trainControl(method = "cv"))
```

If that runs fine, then you could try this with parallel processing.

If you can find the smallest example that exhibits your example, that would be really helpful. You may find the ideas from [reprex](https://reprex.tidyverse.org/), especially [Reprex do's and don'ts](https://reprex.tidyverse.org/articles/reprex-dos-and-donts.html), helpful. Thanks! 🙌

**Comment by DmdCalvinChen** [comment]
@juliasilge Thank you for your reply and efforts. Following your suggestion, I copied the information from the R4.4.0 kernel during the lagging state and named it "R4.4.0Kernel.txt." The other file contains the kernel information of the code you provided, which is running well.

[R4.4.0_YOU_example_code.txt](https://github.com/user-attachments/files/18650345/R4.4.0_You_example_code.txt)

[R4.4.0Kernal.txt](https://github.com/user-attachments/files/18650346/R4.4.0Kernal.txt)

**Comment by DmdCalvinChen** [comment]
@juliasilge Does this log reveal the cause of the problem? If not, I'm planning to build an open-source repository and use a workspace to try reproducing the issue. That way, if it's reproduced, you can get the necessary information from the repository.

**Comment by juliasilge** [comment]
Thanks! Can you help us find more specifics about when you are observing this problem? 

- Do you see the problem when you run a simple caret example with parallel processing? 
- Do you see the problem when you run the neural net example but _without_ parallel processing? 
- Do you see the problem when you run a simpler neural net example (for example, with `iris` or `mtcars`)?

**Comment by DmdCalvinChen** [comment]
@juliasilge Please don't close this issue, I need a few days of free time to study this issue

**Comment by juliasilge** [comment]
The logs make me think maybe ark is having a hard time managing the parallelization maybe?

**Comment by ZhimingYe** [comment]
I have also encountered similar issues during some data analysis work. The core code for this function is as follows:

https://github.com/dmcable/spacexr/blob/0a0861e3d1e16014a20e9b743d0e19d3b42231f3/R/runRCTD.R#L69-L86

As seen, `parallel` is also used. I noticed that under the same conditions, Positron takes significantly longer than RStudio in each loop (sometimes up to 10+ minutes), and surprisingly, the memory usage is about 50GB higher than RStudio. I still need some time to create a reproducible demo. But indeed, there are some strange problems with `parallel` in Positron.

**Comment by DavisVaughan** [comment]
`registerDoParallel()` uses multicore (forked) processing on mac and linux by default. This is currently at odds with Ark's design and is unlikely to work correctly at this time. See https://github.com/posit-dev/positron/issues/3817 for more about this, in particular https://github.com/posit-dev/positron/issues/3817#issuecomment-2207188413.

**Comment by juliasilge** [comment]
Reported again in https://github.com/posit-dev/positron/issues/8460

**Comment by berniejmulvey** [comment]
I just ran into this issue and noticed something interesting under R 4.5.2 -- I can use  Bioconductor 3.22's `bplapply` with `MulticoreParam` in Positron, which itself runs forked processing--no less, it _sets up forked processes using `parallel` behind the scenes_ (and this does seem to succeed: three ark processes show up in `top` while running). The same one-function call within `mclapply` returned NULLs (not even error messages). So, there might be some lead there on how to get forking and Positron to play nice(r)!

Key setup data follows.

•Positron specs: Positron Version: 2026.03.0 build 212
Code - OSS Version: 1.108.0
Commit: f3aae65e0a1a11d39226cd884520f49301daef82
Date: 2026-02-26T04:57:06.999Z
Electron: 39.2.7
Chromium: 142.0.7444.235
Node.js: 22.21.1
V8: 14.2.231.21-electron.0
OS: Darwin arm64 25.3.0

•Mac OS M2 Max running Mac OS 26.3.1
•R 4.5.2
•.Renviron and Renviron.site both include the following line due to forking issues in the past, not sure either of these are still necessary but: `OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES`, sorta following from this post: https://stackoverflow.com/questions/73638290/python-on-mac-is-it-safe-to-set-objc-disable-initialize-fork-safety-yes-globall

•In the script in question I load parallelization libraries and include more historical belt-and-suspenders forking stuff as follows: 
```
library(parallelly)
options(parallelly.supportsMulticore.disableOn="")
options(parallelly.fork.enable=TRUE)
library(BiocParallel)
options(bphost="localhost")
library(parallel)
```

Here's hoping this random observation is helpful for implementing `mcxapply` functionality in Positron!

**Comment by handelsmeisterei** [comment]
I developed a workaround for multicore using foreach that worked till version 2026.08. Works only if your forked processes do basic stuff: 

	result <- foreach(
		i = seq(1, length(prices), by = 1),
		.combine = c,
		.inorder = TRUE,
		.errorhandling = "stop",
		.options.multicore = list(preschedule = FALSE)
	) %dopar%
		#) %do%
		{
			## fix for using %dopar% with positron
			if (!exists(".worker_pid", envir = .GlobalEnv) || get(".worker_pid", envir = .GlobalEnv) != Sys.getpid()) {
				assign(".worker_pid", Sys.getpid(), envir = .GlobalEnv)
				output_file <- tempfile()
				message_file <- tempfile()
				output_con <- file(output_file, "wt")
				message_con <- file(message_file, "wt")
				sink(output_con)
				sink(message_con, type = "message")
				assign(".positron_sink_initialized", TRUE, envir = environment())
				reg.finalizer(
					environment(),
					function(e) {
						sink(type = "message")
						sink()
						close(message_con)
						close(output_con)
					},
					onexit = TRUE
				)
			}

... parallel function ...

Unfortunately 2026.08 explicitly forbids forking now. Would be nice to only have a warning or setting that in the options.

I will am stuck on 2026.07 for the time being because the programs I use profit immensely from forking. Future.apply is a magnitude slower in that case because of copying data to the workers.

**Comment by HUI950319** [comment]
Thank you for keeping this issue open. I would like to add a concrete WSL / remote Linux use case and ask about the roadmap.

Environment:
- Positron 2026.08.1
- WSL 2, Ubuntu 22.04
- R 4.4.3

In our R package workflow, we run repeated Cox and survival-model fits across subgroups. An internal call such as pbapply::pblapply(X, FUN, cl = ncores) selects mclapply/fork on Unix, and Positron 2026.08 now stops immediately with the informative "Can't fork the R session in Positron" error.

Fork-based parallelism is materially useful in WSL and Remote SSH workflows. Large read-only datasets can be shared with workers through copy-on-write memory. PSOCK clusters and multisession workers must start fresh R processes and serialize/export the data and model state to each worker. For large survival, single-cell, and machine-learning datasets, this can substantially increase startup time, network/filesystem traffic, and total memory use; on memory-constrained shared servers, multiple full copies can eliminate the practical benefit of parallelism. WSL and remote Linux servers are also common targets precisely because these workloads are too large or too Linux-dependent for local Windows R.

I understand the safety concerns around forking Ark and why the current release fails early rather than risking hangs or corrupted results. Could the maintainers clarify:

1. Is safe fork support still a long-term goal, or is it currently considered out of scope?
2. If full support is not planned, could an explicitly opt-in advanced mode be considered, with clear warnings and restrictions on frontend/console interaction from workers?
3. Could Positron expose a documented capability flag that packages can query before selecting a backend (for example, something that parallelly::supportsMulticore() and package authors can reliably detect), so packages can choose PSOCK or sequential execution before reaching mcfork()?

We can adapt our own package to use PSOCK in Positron, but many ecosystem packages choose fork automatically on Unix. Package-by-package workarounds therefore do not fully address WSL and Remote SSH users.

Thank you for considering this use case.

**Comment by jmcphers** [comment]
Reported again in https://github.com/posit-dev/positron/issues/15812.
